### DISCUSSION TITLE
Projections

### DISCUSSION THREAD

**Discussion by rhys-rant** [realm/realm-cocoa#7605] [General]
Hi all!

Realm Swift 10.21.0 includes the new `Projection` feature. I'm just a bit curious as to the main purpose of this.
For the given `Person` example, is there any advantage to using a `PersonProjection` instead of creating my own `ViewModel` type? i.e.

```swift
struct PersonViewModel {
  let firstName: String
  let homeCity: String
  let firstFriendsName: [String]

  init(_ person: Person) {
    self.firstName = person.firstName
    self.homeCity = person.address.city
    self.firstFriendsName = person.friends.map(\.firstName)
  }
}
```

Does it optimise queries deep in Realm, or is the feature more about creating a convenient re-usable projection type for use in View Models?

Thanks!

**Comment by astigsen** [comment]
The main benefit is that ViewModels build on Projections stay observable. So you can bind their properties directly in SwiftUI, and the views will automatically update when anything changes.

If you manually build the ViewModel by copying the individual values, you will also have to observe all of them and manually pass on the change events to the UI as they happen. Projections saves you all that work. Think of them as a ViewModel construction kit 😉.
